### PR TITLE
Support ruby 3 with new wrapper interface

### PR DESCRIPTION
**Pull request by Goltergaul**
This redesigns the wrapper interface in order to make it implementable without complications in ruby 3. Additionally I removed any complicated caching logic that was there previously to make the code as simple as possible. For the performance I'm not concerned, as these repository methods are not something that are called quite often usually. So I'd expect that this should do just fine.

I discarded THKs way of defining the wrappers by using rubys module `prepend` mechanism because I wanted to keep the existing gem interface. This should make it much easier to upgrade from 1.0 to 2.0 imo.

There are only a few specs so far, please let me know if there is anything else that should be added (maybe a countdown latch test to make sure it is thread safe?)

This does not yet contain changes we need to do to transfer this to the runtastic org. This will be separate follow up thing

This PR addresses issue #12

### CODE REVIEW

**Comment by Goltergaul** [comment]
add upgrade instructions

**Review by TheWudu — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -225,47 +225,28 @@ applying them in the business logic by using wrappers.
...
+methods named like the method that the repository/strategy method this action should be applied to.
 
 ```ruby
 module Wrapper
   class Validator
-    def before_find(id:)
+    def find(id:)
       raise ArgumentError if id.nil?
-      {id: id}
+      vield(id: id)
```
Comment (ThomasKoppensteiner) [inline README.md:240]: `yield`

File: lib/receptacle/version.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Receptacle
-  VERSION = "1.0.0"
+  VERSION = "1.1.0"
```
Comment (ThomasKoppensteiner) [inline lib/receptacle/version.rb:4]: `2.0.0` ?

File: receptacle.gemspec [hunk 3]
```diff
@@ -27,8 +27,13 @@ Gem::Specification.new do |spec|
...
   spec.add_development_dependency "rake", "~> 10.0"
-  spec.add_development_dependency "standard"
+  spec.add_development_dependency "rspec", "~> 3.11"
+  spec.add_development_dependency "rt_rubocop_defaults", "~> 2.4"
+  spec.add_development_dependency "rubocop", "~> 1.37"
+  spec.add_development_dependency "rubocop-rspec", "~> 2.14"
+  spec.add_development_dependency "rubocop_runner", "~> 2.2"
   spec.add_development_dependency "simplecov", "~> 0.13"
+  spec.add_development_dependency "standard"
+  spec.metadata["rubygems_mfa_required"] = "true"
```
Comment (ThomasKoppensteiner) [inline receptacle.gemspec:38]: I would move this up to line 17 or so.

File: upgrade_notes.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      raise ArgumentError if id.nil?
+      yield(id: id)
+    end
+  end
+  
+  class ModelMapper
+    def find(id:)
+      Model::User.new(yield(id: id))
+    end
+  end
```
Comment (ThomasKoppensteiner) [inline upgrade_notes.md:40]: To make this example a bit more clear I would explicitly save the `return_value`.

```ruby
class ModelMapper
    def find(id:)
      # before: do nothing
      return_value = yield(id: id)
      Model::User.new(return_value) # after
    end
  end
```

File: upgrade_notes.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  class ModelMapper
+    def after_find(return_value, **_kwargs)
+      Model::User.new(return_value)
+    end
+  end
+end
+```
+
+In version 2.0 you need to name the wrapper methods exactly as they are named in your strategy and you call `yield` to call the next wrapper. E.g:
+
```
Comment (ThomasKoppensteiner) [inline upgrade_notes.md:26]: I would add, that wrapper methods follow the around-pattern.

File: lib/receptacle/version.rb (hunk 2 above)
Comment (Goltergaul) [reply lib/receptacle/version.rb:4]: ups yes

File: upgrade_notes.md (hunk 4 above)
Comment (erikpaperik) [reply upgrade_notes.md:40]: Would also be applicable to the example in the README

File: spec/receptacle_spec.rb [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  
+      it "calls the method on the strategy" do
+        expect(repository.find(10, unscoped: false)).to eql(60)
+  
+        expect(strategy_instance).to have_received(:find)
+          .with(10, unscoped: false)
+      end
+    end
+  
+    context "with two wrappers that modify input and output of the find method" do
```
Comment (erikpaperik) [inline spec/receptacle_spec.rb:56]: Within this context, we have `describe` blocks for `find` and `deleted`.
The context only mentions `find method` though.

In general, I think we could add an example which implements the former `after_` wrapper. Both provided wrappers belong to the `before_` mechanism if I got it right.

Comment (Goltergaul) [reply spec/receptacle_spec.rb:56]: the add10_flip_unscoped_and_multiply_result_wrapper has both before and after logic. the after part would be the multiplication with 100

**Comment by erikpaperik** [comment]
Since you mention the thread safeness aspect in the description, I think it would be great to have a test which ensures that.

**Review by erikpaperik — APPROVED** [review]

File: spec/receptacle_spec.rb (hunk 6 above)
Comment (erikpaperik) [reply spec/receptacle_spec.rb:56]: Ah true that!

**Review by ThomasKoppensteiner — APPROVED** [review]

File: spec/receptacle_spec.rb (hunk 6 above)
Comment (erikpaperik) [reply spec/receptacle_spec.rb:67]: This part is clear.
I was confused by the the way we nest the contexts.

**Comment by Goltergaul** [comment]
@andreaseger please also have a look and if it's fine from your side please merge and release. As for testing: I used receptacle 2 already for our workout service and could get the whole test suite running with this code. You can find a pr in stash if you want to have a look. If you want we could also release a RC version first and deploy a service with it?

**Comment by andreaseger** [comment]
> @andreaseger please also have a look and if it's fine from your side please merge and release. As for testing: I used receptacle 2 already for our workout service and could get the whole test suite running with this code. You can find a pr in stash if you want to have a look. If you want we could also release a RC version first and deploy a service with it?

Before any proper review here we need to clarify ownership and maintenance. As mentioned internally I don't have time to maintain this further especially with bigger changes as proposed here. So we need to clarify this and move the repo first before we/you can make bigger changes.

**Pull request closed without merging**